### PR TITLE
Move Persistent Queue to exporterhelper/internal

### PR DESCRIPTION
**Pull request by pmm-sumo**
Description:
This addresses one of the items raised in #4025 (move persistent_queue to internal). The refactor impacted several things:

* Some names had to be exported - e.g. `internal.RequestUnmarshaler` or `internal.PersistentRequest` 
* I have extracted `BoundedQueue` API as an interface. The memory-based implementation is now known as `boundedMemoryQueue`

Link to tracking Issue: #4025

Testing: Unit tests updated, manual tests to follow

### CODE REVIEW

**Comment by bogdandrutu** [comment]
@pmm-sumo we discussed that this needs to be split into separate PRs, is this ready for review?

**Comment by pmm-sumo** [comment]
Hi Bogdan, I am on vacations until the end of this week but splitting the
PR will be first thing when I get back on Monday. Apologies for the delay

W dniu pt., 24.09.2021 o 13:41 Bogdan Drutu ***@***.***>
napisał(a):

File: exporter/exporterhelper/consumers_queue.go [hunk 1]
```diff
@@ -14,10 +14,10 @@
 
 package exporterhelper
 
-// consumersQueue is largely based on queue.BoundedQueue and matches the subset used in the collector
+// ConsumersQueue is largely based on queue.BoundedQueue and matches the subset used in the collector
 // It describes a producer-consumer exchange which can be backed by e.g. the memory-based ring buffer queue
 // (queue.BoundedQueue) or via disk-based queue (persistentQueue)
-type consumersQueue interface {
+type ConsumersQueue interface {
```
Comment (tigrannajaryan) [inline exporter/exporterhelper/consumers_queue.go:20]: Does this need to be exporter? I cannot see it referenced anywhere.

File: exporter/exporterhelper/factory.go [hunk 2]
```diff
@@ -78,6 +79,7 @@ func NewFactory(
 	for _, opt := range options {
 		opt(f)
 	}
+	internal.RegisterMetrics()
```
Comment (tigrannajaryan) [inline exporter/exporterhelper/factory.go:82]: Can you please clarify why this change? Does this add a functionality or it is forced by refactoring?

File: exporter/exporterhelper/internal/observability_experimental.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+	"go.opencensus.io/tag"
+)
+
+var (
+	currentlyDispatchedBatches = stats.Int64(
+		"/currently_dispatched_batches",
+		"Number of batches that are currently being sent",
+		stats.UnitDimensionless)
+
+	totalDispatchedBatches = stats.Int64(
```
Comment (tigrannajaryan) [inline exporter/exporterhelper/internal/observability_experimental.go:35]: This seems to add a new metric to unstable. Can we do it in a separate PR to keep the refactoring clearer?

File: exporter/exporterhelper/consumers_queue.go (hunk 1 above)
Comment (pmm-sumo) [reply exporter/exporterhelper/consumers_queue.go:20]: That's true, I overlooked it. I will remove it

**Comment by pmm-sumo** [comment]
Closing this PR in favor of three smaller ones, the first one being: https://github.com/open-telemetry/opentelemetry-collector/pull/4115
